### PR TITLE
kafkabp: Support AWS IMDS v2 for rack id

### PR DESCRIPTION
**Pull request by fishy**
Also modernize the code a bit (use sync.OnceValues, etc.).

### CODE REVIEW

File: kafkabp/rack.go [hunk 1]
```diff
@@ -126,55 +127,104 @@ func SimpleHTTPRackID(cfg SimpleHTTPRackIDConfig) RackIDFunc {
...
 // Global cache for AWSAvailabilityZoneRackID.
-var (
-	awsCachedRackID string
-	awsRackIDOnce   sync.Once
-)
+var awsRackID = sync.OnceValues(func() (string, error) {
+	const (
+		// References:
+		// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
+		// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
```
Comment (kylelemons) [inline kafkabp/rack.go:189]: 🔕 is there a reference for the IMDSv2 / token API to add?

File: kafkabp/rack.go [hunk 2]
```diff
@@ -126,55 +127,104 @@ func SimpleHTTPRackID(cfg SimpleHTTPRackIDConfig) RackIDFunc {
...
+		tokenURL = "http://169.254.169.254/latest/api/token"
+		azURL    = "http://169.254.169.254/latest/meta-data/placement/availability-zone"
+
+		timeout   = time.Second
```
Comment (kylelemons) [inline kafkabp/rack.go:193]: 🔕 I think units should always come with a scalar too for readability, even if it's `1`

```suggestion
		timeout   = 1*time.Second
```

File: kafkabp/rack.go [hunk 3]
```diff
@@ -126,55 +127,104 @@ func SimpleHTTPRackID(cfg SimpleHTTPRackIDConfig) RackIDFunc {
...
+		readLimit = 1024
+	)
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	token, err := func(ctx context.Context) (string, error) {
```
Comment (kylelemons) [inline kafkabp/rack.go:200]: Same comments as the `rack` API: not sure I am seeing the reason for the anonymous function, and that header should be normalized

File: kafkabp/rack.go [hunk 4]
```diff
@@ -194,18 +244,18 @@ const awsAZurl = "http://169.254.169.254/latest/meta-data/placement/availability
...
-				nil, // delegate, let it fallback to DefaultWrapper
-				awsRackFailure,
-			),
-		})()
-	})
-	return awsCachedRackID
+	id, err := awsRackID()
+	if err != nil {
+		awsRackFailure.Inc()
+		slog.Warn("Failed to get AWS availability zone as rack id", "err", err)
```
Comment (kylelemons) [inline kafkabp/rack.go:257]: 🔕 is v0 ready for slog?

**Review by kylelemons — APPROVED** [review]

File: kafkabp/rack.go (hunk 4 above)
Comment (fishy) [reply kafkabp/rack.go:257]: I think we want to gradually shift v0 to use slog directly, which will use ctxlog if services have ctxlog set up in their main (as v0 cannot use ctxlog directly)

**Pull request merged**